### ISSUE TITLE
Modify schema with a closure

### ISSUE THREAD

**Issue by IlyaSemenov** [colinhacks/zod#2281]
### Problem

From what I see, that is not possible to construct a zod schema conditionally without `let`, temporary local variables, self-executing function and such.

Let's say I have a schema for a user and their address:

```ts
const user_schema = z.object({
  name: z.string(),
  address: z.object({
    city: z.string(),
  }),
})
```

Now let's imagine I create this schema in a context, and based on the context I may want city to be nullable:

```ts
function create_user_schema(ctx: { city_nullable: boolean }) {
  let city_schema = z.string()
  if (ctx.city_nullable) {
    city_schema = city_schema.nullable()
  }

  return z.object({
    name: z.string(),
    address: z.object({
      city: city_schema,
    }),
  })
}
```

Not only this is ugly and lengthy: with duck typing as above, this leads to type error: `Type 'ZodNullable<ZodString>' is missing the following properties from type 'ZodString': _regex, _addCheck, email, url, and 29 more` (I realise that the error could be avoided by an additional intermediate schema variable - that's not the point).

As the logic extends into the tree of variants, this type of code becomes hard to maintain.

### Proposal

Inspired by [Knex modify](https://knexjs.org/guide/query-builder.html#modify), I propose to add a new generic method to `ZodType`, so that the example above could be converted to:

```ts
function create_user_schema(ctx: { city_nullable: boolean }) {
  return z.object({
    name: z.string(),
    address: z.object({
      city: z.string().modify(schema => {
        // Executed once upon schema creation - not when parsing
        return ctx.city_nullable ? schema.nullable() : schema
      }),
    }),
  })
}
```

The code becomes much simpler. The 'modification' is directly attached to the proper position inside the (possibly large) schema tree. The namespace is not polluted needlessly.

If I understand correctly, the actual implementation will be as simple as:

```ts
class ZodType {
  // TODO: add proper typings
  modify(fn) {
    return fn(this)
  }
}
```

**Comment by IlyaSemenov** [comment]
Oh...I really hoped I wouldn't need to explain this. **No, your answer does not help.**

For starters, imagine it's not a `z.string()` but say `z.string().email().refine(...).transform(...)`. Are you going to copy-paste that near the `const foo =`?

Then, these manual typings are ugly and don't really make sense. As I told, the nullable flag comes from a context (such as from database or the HTTP request) - it's simply a boolean. It's not known beforehand if that's true or false. That `Nullable extends true` boilerplate is therefore useless.

All and all, I am proposing a **generic enhancement**. Making a sub-schema nullable was just a simple example. It's far more than that that is needed in real life. For instance, an object schema could be extended to have more fields or to omit certain fields based on context. Refines could be dynamically added. All sort of modification that currently require local top-level variables could be incapsulated into context-bound functions.

**Comment by JacobWeisenburger** [comment]
> Oh...I really hoped I wouldn't need to explain this. **No, your answer does not help.**

This is a bit rude. I'm just trying to help. There is no need to be rude.

**Comment by IlyaSemenov** [comment]
Fair enough, but I was not asking for help. I was proposing API enhancement to the library - that is an issue directed towards library maintainers.

I am an author of similar data validation library ([link](https://github.com/IlyaSemenov/node-data-cleaner/tree/master/packages/data-cleaner)) which I created 3 years before zod even appeared. Now since I'm migrating to zod, I'm seeing certain shortcomings when it comes to ease of use.

I expect greater expertise from the maintainers - that's why I told I was hoping I wouldn't need to explain that this issue is not about me struggling to make a schema nullable.

**Comment by denizdogan** [comment]
All rudeness aside, I am also looking for a clean way to do something like this! +1

**Comment by ankology** [comment]
All rudeness aside², you can manually tell the possible types to `city_schema`, just replace and type inference will work:
```
let city_schema: ZodString | ZodNullable<ZodString> = z.string();
```

**Comment by IlyaSemenov** [comment]
@ankology Try to extend the use case beyond the minimal example:

1) city could be nullable or not
2) if not nullable, it could have a context-based max length (or not!)
3) then it could be optional...or not!

Now then, what if that's not a scalar but `z.object` which has to be slightly different on different nesting levels? That would require a number of temporary variables (and possibly types) and hardly readable boilerplate to compose the final schema.

As I elaborated in the op-post: _"As the logic extends into the tree of variants, this type of code becomes hard to maintain."_

Hope that makes sense.

**Comment by IlyaSemenov** [comment]
Whatever, let it stale. I switched to [valibot](https://github.com/fabian-hiller/valibot), which is composition-based rather than puts everything in the base class, and makes it extremely easy to add own plugins:

```ts
import { type BaseSchema } from "valibot"

export function modify<TSchema extends BaseSchema, TModifiedSchema extends BaseSchema>(
  schema: TSchema,
  modifier: (schema: TSchema) => TModifiedSchema,
) {
  return modifier(schema)
}
```

and then:

```ts
import * as v from "@my/valibot-with-extensions"

function create_user_schema(ctx: { city_nullable: boolean }) {
  return v.object({
    name: v.string(),
    address: v.object({
      city: v.modify(v.string(), (schema) => {
        // Executed once upon schema creation - not when parsing
        return ctx.city_nullable ? v.nullable(schema) : schema
      }),
    }),
  })
}
```

**Comment by rafaell-lycan** [comment]
@IlyaSemenov I bet that in a few months, Zod will evolve and add a few other features that developers are aiming for, or focus on performance.

There's a heck of an ecosystem already using Zod as the fav schema validator over other solutions, and over time this will make other projects such `valibot` bring something to the table and perhaps make Zod do the same. (Like npm vs yarn drama)

I guess that your suggestion is great, and someone from the community or even yourself could contribute to bring it to Zod.

Cheers.

**Comment by mohrazzak** [comment]
It's probably too late but I had something similiar 

My problem: password is srtring or null but in some DTO I want it to be string
I had to do it manually like this 
```ts
export class ConfirmSignupDTO extends createZodDto(
  StoreOwnerModel.pick({
    email: true,
    signupCode: true,
  })
    .extend({ password: z.string() })
    .strict()
    .required(),
) {}
```

**Comment by IlyaSemenov** [comment]
> Otherwise, this issue will be automatically closed in 28 days.

That's funny how 28 days turned into one week. :)

**Comment by colinhacks** [comment]
I'm not really seeing how this proposal is materially different from what you can already achieve with a schema factory pattern + ternaries. It's trivial to rewrite your original "Problem" code block to be just as elegant looking as the code in your proposed solution.
